src/osd/ReplicatedPG.cc

```diff
@@ -1746,6 +1746,14 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // test the ondisk_read_lock as early as possible to avoid stuck at a later time
+  if (op->may_read() && !obc->test_ondisk_read_lock()) {
+    dout(20) << __func__ << " waiting for ondisk_read_lock " << dendl;
+    op->mark_delayed("waiting for ondisk_read_lock");
+    close_op_ctx(ctx, -EBUSY);
+    return;
```

Comment:
This will just leak the request since we aren't putting it on any list for later processing.  That means we'd need a wait list and we need to requeue the op when the lock is dropped.

Note that even if we did, we should make sure that we are not write_ordered or else the ordering constraint is broken. 

If there is an opportunity for requeue that is pretty simple this *might* work, but I suspect we need a more robust async read framework to make this really work properly..